File: docs/agents/browserbase/node/playwright/_tools/stealth/customCaptcha.ts
/**
 * Run using npx tsx captcha/custom-captcha.ts
**/

import { chromium } from "playwright-core";
import Browserbase from "@browserbasehq/sdk";
import dotenv from "dotenv";
dotenv.config();

const bb = new Browserbase({apiKey: process.env.BROWSERBASE_API_KEY!});

async function createCustomCaptchaSession() {
    // Create a new session
    const session = await bb.sessions.create({
      projectId: process.env.BROWSERBASE_PROJECT_ID!,
      browserSettings: {
        //@ts-ignore
          captchaImageSelector: "#c_turingtestpage_ctl00_maincontent_captcha1_CaptchaImage",
          captchaInputSelector: "#ctl00_MainContent_txtTuringText"
        }
    });

    // Connect to the session
    return session;
}

async function runCustomCaptchaSession() {
  const session = await createCustomCaptchaSession();
  const browser = await chromium.connectOverCDP(session.connectUrl);

  // Getting the default context to ensure the sessions are recorded.
  const defaultContext = browser.contexts()[0];
  const page = defaultContext?.pages()[0];
  console.log(
  `View sessionreplay at https://browserbase.com/sessions/${session.id}`,
  );
  // Navigate to website with custom captcha
  const consumerAccess = await page.goto("https://nmlsconsumeraccess.org/");
  await page.waitForLoadState('networkidle');

  // Search for company
  await page.waitForSelector('input[type="text"]');
  await page.click('input[type="text"]');
  await page.locator('input[type="text"]').first().fill(company);

  // Click search
  await page.click('input[id="searchButton"]');

  // Click the terms and conditions checkbox
  await page.click('#ctl00_MainContent_cbxAgreeToTerms');

  // wait for page to load
  console.log("Captcha auto-solving...");
  await page.waitForTimeout(5000); // can add this - but CAPTCHA is usually solved much quicker

  await page.click('#ctl00_MainContent_btnContinue');
  await page.waitForLoadState('networkidle');

  console.log("Shutting down...");
  await page.close();
  await browser.close();
}

const company = "alt platform";
runCustomCaptchaSession();

File: docs/research/agents/browserbase/python/playwright/captcha/custom_captcha.py
from playwright.sync_api import Playwright, sync_playwright
from browserbase import Browserbase
import os

company = "alt platform";

bb = Browserbase(api_key=os.environ["BROWSERBASE_API_KEY"])

def run(playwright: Playwright) -> None:
    session = bb.sessions.create(
        project_id=os.environ["BROWSERBASE_PROJECT_ID"],
        browser_settings={
            "captchaImageSelector": "#c_turingtestpage_ctl00_maincontent_captcha1_CaptchaImage",
            "captchaInputSelector": "#ctl00_MainContent_txtTuringText"
        }
    )
    # Connect to the remote session
    browser = playwright.chromium.connect_over_cdp(session.connect_url)
    context = browser.contexts[0]
    page = context.pages[0]

    print(f"View session replay at https://browserbase.com/sessions/{session.id}")
    
    # Navigate to the website
    page.goto("https://nmlsconsumeraccess.org/")
    
    # Wait for page load
    page.wait_for_load_state("networkidle")
    print("Success!")
    
    # Search for company
    page.wait_for_selector('input[type="text"]')
    page.click('input[type="text"]')
    page.locator('input[type="text"]').first.fill(company)
    
    # Click search
    page.click('input[id="searchButton"]')
    
    # Click the terms and conditions checkbox
    page.click('#ctl00_MainContent_cbxAgreeToTerms')
    
    # Wait for captcha to be solved
    print("Captcha auto-solving...")
    page.wait_for_timeout(5000)
    
    # Click continue
    page.click('#ctl00_MainContent_btnContinue')
    page.wait_for_load_state("networkidle")
    
    print("Shutting down...")
    page.close()
    browser.close()

if __name__ == "__main__":
    with sync_playwright() as playwright:
        run(playwright)
